scipy/sparse/linalg/_krylov_funmv.py

```diff
@@ -0,0 +1,266 @@
+
+def _krylov_funmv_lanczos(A, b, bnorm, V, H, m):
+    """
+    The Lanczos iteration for constructing the basis V and the projection H = V * A V
+    for the Krylov subspace Km(A, b) of order m. A must be Hermitian.
+
+    Parameters
+    ----------
+    A : transposable linear operator
+        The operator whose matrix function is of interest.
+    b : ndarray
+        The vector b to multiply the f(A) with.
+    V : ndarray
+        The n x (m + 1) matrix whose columns determines the basis for
+        Krylov subspace Km(A, b).
+    H : ndarray
+        A (m + 1) x m upper Hessenberg matrix representing the projection of A
+        onto Km(A, b).
+    m : int
+        The order of the Krylov subspace.
+
+    """
+    dotprod = np.vdot if np.iscomplexobj(b) else np.dot
+    V[:, 0] = b / bnorm
+
+    for k in range(0, m):
+        if k > 0:
+            V[:, k + 1] = A.dot(V[:, k]) - H[k, k - 1] * V[:, k - 1]
+        else:
+            V[:, k + 1] = A.dot(V[:, k])
+
+        H[k, k] = dotprod(V[:, k + 1], V[:, k])
+        V[:, k + 1] = V[:, k + 1] - H[k, k] * V[:, k]
+
+        H[k + 1, k] = np.linalg.norm(V[:, k + 1])
+        if H[k + 1, k] != 0:
+            V[:, k + 1] = V[:, k + 1] / H[k + 1, k]
+            if k < m - 1:
+                H[k, k + 1] = H[k + 1, k]
+        else:
+            raise RuntimeError("krylov_funmv: The Lanczos iteration broke down "
+                               f"at k = {k}, i.e., V[:, k + 1] = 0")
+
+
+def krylov_funmv(f, t, A, b, *, atol = 0.0, btol = 1e-6, restart_length = None,
+                 max_restarts = 20, ortho_method = "arnoldi", verbose = False):
+    """
+    A restarted Krylov method for evaluating ``y = f(tA) b`` from [1]_ [2]_.
+
+    Parameters
+    ----------
+    f : function
+        Callable object that computes the matrix function ``F = f(X)``.
+
+    t : float
+        The value to scale the matrix ``A`` with.
+
+    A : {sparse array, ndarray, LinearOperator}
+        A real or complex N-by-N matrix.
+        Alternatively, `A` can be a linear operator which can
+        produce ``Ax`` using, e.g., ``scipy.sparse.linalg.LinearOperator``.
+
+    b : ndarray, sparse array
+        A vector to multiply the ``f(tA)`` with.
+
+    atol, btol : float, optional
+        Parameters for the convergence test. For convergence,
+        ``norm(||y_k - y_k-1||) <= max(btol*norm(b), atol)`` should be satisfied.
+        The default is ``atol=0.`` and ``btol=1e-6``.
+
+    restart_length : integer
+        Number of iterations between restarts. Larger values increase
+        iteration cost, but may be necessary for convergence.
+        If omitted, ``min(20, n)`` is used.
+
+    max_restarts : int, optional
+        Maximum number of restart cycles.  The algorithm will stop
+        after max_restarts cycles even if the specified tolerance has not been
+        achieved. The default is ``max_restarts=20``
+
+    ortho_method : string, optional
```

Comment:
In the linalg module we use the keyword `assume_a` with values `gen` and `her` for this (see `linalg.solve` for example). If it is `her`, it can dispatch to Lanczos otherwise does the Arnoldi. Probably should go first in the order of keywords.

Reply:
Ah yes, it does not need to detect automatically but a user can hint that the operator is Hermitian. So we select the method based on that.